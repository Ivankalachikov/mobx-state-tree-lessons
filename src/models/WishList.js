import { types } from 'mobx-state-tree';

// const data = {
//   name: 'Some amazing book',
//   price: 1500,
//   image: '//via.placeholder.com/350x150'
// }

export const WishListItem = types.model({
  name: types.string,
  price: types.number,
  image: '',
})

export const WishList = types.model({ 
  items: types.optional(types.array(WishListItem), []),
});