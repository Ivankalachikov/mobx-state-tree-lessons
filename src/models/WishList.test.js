import { getSnapshot, onPatch, onSnapshot } from 'mobx-state-tree';
import { WishList, WishListItem } from './WishList';

test('Can create an instance of a model', () => {
  const item = WishListItem.create({
    name: 'Some amazing book',
    price: 1500,
  })
  expect(item.price).toBe(1500);
  expect(item.image).toBe('');
  item.changeName('Book');
  expect(item.name).toBe('Book');
})

test('Can create a wishlist', () => {
  const list = WishList.create({
    items: [{
      name: 'Some amazing book',
      price: 1500,
    }]
  });
  expect(list.items.length).toBe(1);
  expect(list.items[0].price).toBe(1500);
})

test('can add new items - snapshots', () => {
  const states = [];

  const list = WishList.create();
  onSnapshot(list, (snapshot) => {
    states.push(snapshot);
  })

  list.add({ name: 'Socks', price: 100 });

  expect(list.items.length).toBe(1);
  expect(list.items[0].name).toBe('Socks');
  
  list.items[0].changeName('Flowers');
  expect(list.items[0].name).toBe('Flowers');
  expect(getSnapshot(list)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
})

test('can add new items - patches', () => {
  const patches = [];

  const list = WishList.create();
  onPatch(list, (patch) => {
    patches.push(patch);
  })

  list.add({ name: 'Socks', price: 100 });
  list.items[0].changeName('Flowers');

  expect(patches).toMatchSnapshot();
})