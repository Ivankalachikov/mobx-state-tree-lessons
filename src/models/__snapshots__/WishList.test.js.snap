// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items - patches 1`] = `
[
  {
    "op": "add",
    "path": "/items/0",
    "value": {
      "image": "",
      "name": "Socks",
      "price": 100,
    },
  },
  {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Flowers",
  },
]
`;

exports[`can add new items - snapshots 1`] = `
{
  "items": [
    {
      "image": "",
      "name": "Flowers",
      "price": 100,
    },
  ],
}
`;

exports[`can add new items - snapshots 2`] = `
[
  {
    "items": [
      {
        "image": "",
        "name": "Socks",
        "price": 100,
      },
    ],
  },
  {
    "items": [
      {
        "image": "",
        "name": "Flowers",
        "price": 100,
      },
    ],
  },
]
`;
